### DISCUSSION TITLE
Beacon request cancelation missing on page navigation in Cypress

### DISCUSSION THREAD

**Discussion by mduchemin** [cypress-io/cypress#30129] [Questions and Help]
We are encountering a situation where the testing environment causes a difference in behavior.
Environment: Chrome for Windows, v 127.0.6533.120

## Context:
- The application tested is servlet based, with Spring-Security authentication
- When a request is made to the webserver, if there is currently no active session or if the session is invalidated, a new server-side session will be created. When this happens, the response will contain a cookie header setting the JSessionId cookie to the client.
- The application sends a [beacon ](https://developer.mozilla.org/en-US/docs/Web/API/Beacon_API) request to a dedicated endpoint on the server during window beforeUnload. This request is Async and non-blocking, so other browser behaviors continue while the request / response cycle occurs. This request is NOT expected to provide a response to the client, since beforeUnload triggers when the page is being destroyed (navigation, tab closing, etc.). Instead, its purpose is to notify the server that the client is unloading in order to perform server-side cleanup without waiting for session expiration. Since beacon is async and non-blocking, this allow the server to receive the request even if the emiting document has already unloaded by the time the request reaches the webserver, as the beacon request is not dependant on the source document.
- the beacon response may contain a setCookie header for JSessionId, if the previous session was invalidated or expired. Spring-security will automatically invalidate sessions during some operations, such as login with form. If navigation occurs just after session invalidation such as during login, both the beacon request and document request are sent holding the already invalidated JSessionID, so each separately create a new servlet session and their response contain each a different JSessionId cookie header.

### In a regular browser

![beacon-unload drawio](https://github.com/user-attachments/assets/8ef0b2f6-9e6d-455b-9ff6-b790c5eb591e)

The beacon request is sent when navigation to another page is requested (for example, by changing window.location).
This means that there are 2 competing requests emitted at the same time: Navigation document request and beacon notification request.

If the beacon request completes first, the document will receive the response first, setCookie, then receive the document response containing its own setCookie. In this case, the document's cookie is the final one because it overwrite the beacon cookie.

If the document request completes first, the beacon request is dropped by the browser. This doesn't prevent the request from completing its server-side purpose, but any response is discarded (including headers). In this case, the document's cookie is the final one because the beacon cookie is never set.

### In Cypress testing environment
![beacon-unload drawio (1)](https://github.com/user-attachments/assets/6d3c4fc9-b66a-4f80-bb4e-6ab00e93bea4)

The difference is caused by the fact that the cypress environment doesn't actually unload and navigate away from the tested document. My understanding is that in the navigator tab, Cypress "emulates" the request / response transactions, and use the content to change the testing area. Since the beacon request is tied to Navigator, the request is not discarded by the browser as it 

If the beacon request completes first, the document will receive the response first, setCookie, then receive the document response containing its own setCookie. In this case, the document's cookie is the final one because it overwrite the beacon cookie. This case is identical to the standard browser use case.

If the document request completes first, the beacon request is NOT dropped by the browser since the actual browser window hosting the cypress test suite is still the same context. As a result, the browser will perform the document's request setCookie, followed by the beacon's setCookie, which overwrites it.
As a result, the page will load the document, but use a different JSessionId cookie which doesn't match the document's context.

## Questions:

- Is there any configuration switch to set cypress to discard responses to beacon requests emitted by a tested page which has already been unloaded?
- Is this intended in Cypress? If this is not up to current spec, should I create a bug or feature in the issues tracker?

**Comment by portyu9** [comment]
Hi @mduchemin 

I would fix this on the server rather than depending on Cypress to cancel/discard the beacon response.

`navigator.sendBeacon()` is specifically designed for requests that need to survive page unload. Its underlying fetch uses `keepalive`, so there is **not a reliable browser contract that the response will be discarded simply because the old document has unloaded**.

That means this sequence is inherently racy if the beacon endpoint can create a new session:

1. Login/document navigation uses the old/invalidated session.
2. Cleanup beacon also uses that session.
3. Both responses are capable of sending a new `JSESSIONID`.
4. Whichever cookie update wins can determine the final session.

A cleanup beacon should therefore ideally be **stateless with respect to authentication/session creation**. For example, the endpoint should not create or rotate a `JSESSIONID` when the old session is already invalid, and can return something like `204 No Content` without changing the authentication cookie.

I am not aware of a supported Cypress configuration option whose contract is “discard beacon responses after the originating page unloads”, and relying on such behavior would make the application dependent on a browser/network race anyway.

There is also an important current-version difference: starting with Cypress 16, Chrome/Chromium/Edge use Cypress's newer native browser networking path, where the application connects to the server through the browser rather than the historical Cypress proxy path. So I would re-test this reproduction on Cypress 16 before concluding the 2024 Cypress-specific networking difference still exists.

If it still reproduces on current Cypress Chrome, capture the two response `Set-Cookie` headers and server-side timestamps in both normal Chrome and Cypress and report the minimal reproduction.

I could not retrieve either attached diagram because the old GitHub attachment URLs now return 404, so I did not rely on them.

Please mark this answer as accepted if it helped, thank you.
